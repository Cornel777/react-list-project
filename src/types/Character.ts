export type Gender = 'male' | 'female' | 'n/a' | 'none';
export type Color = string;
export type URL = string;

export interface Character {
  name: string;
  height: string;
  mass: string;
  hair_color: Color;
  skin_color: Color;
  eye_color: Color;
  birth_year: string;
  gender: Gender;
  homeworld: URL;
  films: URL[];
  species: URL[];
  vehicles: URL[];
  starships: URL[];
  created: string;
  edited: string;
  url: URL;
}
