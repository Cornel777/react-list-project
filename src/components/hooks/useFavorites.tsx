import {useState} from 'react'
import { Character } from '../../types/Character';


export const STORAGE_KEY = 'SW_favorites';

interface HookReturn {
  favorites: Character[];
  toggleFavorite: (character: Character) => void;
  isFavorite: (character: Character) => boolean;
}

export const useFavorites = (): HookReturn => {
  const [favorites, setFavorites] = useState<Character[]>(() => {
    try {
      const stored = localStorage.getItem(STORAGE_KEY);
      return stored ? JSON.parse(stored) : [];
    } catch (error) {
      console.error('Error encoutered while fetching favorites list' , error);
      return [];
    }
});

  const toggleFavorite = (character: Character) => {
    try {
      setFavorites((prevFavorites) => {
        const isStored = prevFavorites.some((favorite) => favorite.url === character.url);

        let newFavorites: Character[];
        if (isStored) {
          newFavorites = prevFavorites.filter((favorite) => favorite.url !== character.url);
        } else {
          newFavorites =  [...prevFavorites, character];
        }
        
        localStorage.setItem(STORAGE_KEY, JSON.stringify(newFavorites));
        return newFavorites;
      })
    } catch (error) {
      console.error('Error encountered while toggling favorite', error);
    }
  }

  const isFavorite = ( character: Character) => {
    return favorites.some((favorite) => favorite.url === character.url);
  }
  return { favorites, toggleFavorite, isFavorite };
}

export default useFavorites;