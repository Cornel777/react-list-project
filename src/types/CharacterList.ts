import { Character } from "./Character";
import { URL } from "./Character";

export interface CharacterList {
  count: number;
  next: URL | null;
  previous: URL | null;
  results: Character[];
}