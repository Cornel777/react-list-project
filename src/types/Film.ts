import { URL } from './Character';

export interface Film {
  title: string;
  episode_id: number;
  opening_crawl: string;
  director: string;
  producer: string;
  release_date: string;
  characters: URL[];
  planets: URL[];
  starships: URL[];
  vehicles: URL[];
  species: URL[];
  created: Date;
  edited: Date;
  url: URL;
}