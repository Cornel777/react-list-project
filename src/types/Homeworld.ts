import { URL } from "./Character";

export interface Homeworld {
  name: string;
  rotation_period: string;
  orbital_period: string;
  diameter: string;
  climate: string;
  gravity: string;
  terrain: string;
  surface_water: string;
  population: string;
  residents: string[];
  films: URL[];
  created: string;
  edited: string;
  url: URL;
}

