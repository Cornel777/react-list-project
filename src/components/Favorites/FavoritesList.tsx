import { Typography, TableContainer, Paper, Table, TableHead, TableBody, IconButton, } from "@mui/material";
import DeleteIcon from '@mui/icons-material/Delete';
import { FunctionComponent, useState } from "react";
import Homeworld from "../CharacterDetails/components/Homeworld/Homeworld";
import BreakElement from "../CharacterDetails/components/styled/BreakElement";
import { StyledTableRow, StyledTableCell } from "../CharacterList/styled/styledListTable";
import useFavorites from "../hooks/useFavorites";
import { Character } from "../../types/Character";
import EditableCell from "../EditableCell/EditableCell";


const FavoritesList: FunctionComponent = () => {
  const [isFocused, setIsFocused] = useState<string | null>(null);
  const { favorites, toggleFavorite } = useFavorites();

  const handleToggleFavorite = ( favorite: Character) => { 
    toggleFavorite(favorite)
  };

  const handleKeyDownToggleFavorite = (e: React.KeyboardEvent, favorite: Character) => {
    if (e.key === 'Enter' || e.key === ' ') {
      handleToggleFavorite(favorite)
    }
  };

  const handleFocus = (favoriteId: string) => {
    setIsFocused(favoriteId);
    console.log(favoriteId)
  };

  const handleBlur = () => {
    setIsFocused(null);
  };

  const handleEditSave = (favorite: Character, value: string, field: string) => {
    const updatedCharacter = { ...favorite, [field]: value };
    toggleFavorite(favorite);
    toggleFavorite(updatedCharacter);
  };

  return (
      <div>
        <Typography variant="h3">Favorites List</Typography>
        <BreakElement/>
        <TableContainer component={Paper} style={{ maxWidth: '600px' }}>
          <Table size="medium" aria-label="Favorites list table">
            <TableHead>
              <StyledTableRow>
                <StyledTableCell>Name</StyledTableCell>
                <StyledTableCell>Height</StyledTableCell>
                <StyledTableCell>Gender</StyledTableCell>
                <StyledTableCell>Home Planet</StyledTableCell>
                <StyledTableCell>Remove</StyledTableCell>
              </StyledTableRow>
            </TableHead>
            <TableBody>
              {favorites.map((favorite) => {
                const favoriteUrlList = favorite.url.split('/').filter((element: string) => element.length > 0)
                const favoriteId = favoriteUrlList.pop()
                  return (
                      <StyledTableRow 
                        key={favoriteId}
                        onMouseEnter={() => handleFocus(favoriteId!)}
                        onMouseLeave={handleBlur}
                        onFocus={() => handleFocus(favoriteId!)}
                        onBlur={handleBlur}
                        tabIndex={0}
                      >
                        <StyledTableCell>{favorite.name}</StyledTableCell>
                        <EditableCell
                          value={favorite.height}
                          field='height'
                          favorite={favorite}
                          isFocused={isFocused === favoriteId}
                          onSave={handleEditSave}
                        />
                        <StyledTableCell>{favorite.gender}</StyledTableCell>
                        <StyledTableCell><Homeworld homeworldUrl={favorite.homeworld}/></StyledTableCell>
                        <StyledTableCell>
                        <IconButton 
                          aria-label="remove from favorites button"
                          onClick={() => handleToggleFavorite(favorite)}
                          onKeyDown={(e) => handleKeyDownToggleFavorite(e, favorite)}
                        >
                          <DeleteIcon color="error" />
                        </IconButton> 
                        </StyledTableCell>
                      </StyledTableRow>
                  )
              })}
           </TableBody>
          </Table>
       </TableContainer>
       </div>
  )
}

export default FavoritesList;