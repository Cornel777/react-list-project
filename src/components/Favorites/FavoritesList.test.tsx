import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { render, screen, fireEvent } from '@testing-library/react';
import useFavorites from '../hooks/useFavorites';
import { Character } from "../../types/Character";
import FavoritesList from "./FavoritesList";

vi.mock('../hooks/useFavorites');
vi.mock("../CharacterDetails/components/Homeworld/Homeworld");

const mockCharacter = {
  name: 'Luke Skywalker',
  height: '172',
  gender: 'male',
  homeworld: 'https://swapi.dev/api/planets/1/',
  url: 'https://swapi.dev/api/people/1/',
  films: [],
  species: [],
  vehicles: [],
  starships: [],
  created: '',
  edited: '',
  birth_year: '',
  eye_color: '',
  hair_color: '',
  skin_color: '',
  mass: ''
} as Character;

describe('FavoritesList', () => {
  beforeEach(() => {
    vi.mocked(useFavorites).mockReturnValue({
      favorites: [mockCharacter],
      toggleFavorite: vi.fn(),
      isFavorite: vi.fn()
    });
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  it('should render the list of favorites', () => {
    const {container} = render(<FavoritesList />);
    expect(container).toMatchSnapshot();
  });

  it('renders the list of favorites', () => {
    render(<FavoritesList />);
    const name = screen.getByText('Luke Skywalker');
    expect(name).toBeInTheDocument();
    expect(screen.getByText('172')).toBeInTheDocument();
    expect(screen.getByText('male')).toBeInTheDocument();
  });

  it('toggles favorite on click', () => {
    render(<FavoritesList />);
    const removeButton = screen.getByLabelText('remove from favorites button');
    fireEvent.click(removeButton);
    expect(useFavorites().toggleFavorite).toHaveBeenCalledWith(mockCharacter);
  });

  it('toggles favorite on keydown', () => {
    render(<FavoritesList />);
    const removeButton = screen.getByLabelText('remove from favorites button');
    fireEvent.keyDown(removeButton, { key: 'Enter', code: 'Enter' });
    expect(useFavorites().toggleFavorite).toHaveBeenCalledWith(mockCharacter);
  });

});