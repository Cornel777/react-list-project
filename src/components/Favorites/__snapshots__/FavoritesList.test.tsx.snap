// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoritesList > should render the list of favorites 1`] = `
<div>
  <div>
    <h3
      class="MuiTypography-root MuiTypography-h3 css-13msox4-MuiTypography-root"
    >
      Favorites List
    </h3>
    <br
      class="sc-KXCwU"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); max-width: 600px;"
    >
      <table
        aria-label="Favorites list table"
        class="MuiTable-root css-1xwxv7r-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-r0kehe-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Height
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Gender
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Home Planet
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Remove
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-r0kehe-MuiTableRow-root"
            tabindex="0"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              Luke Skywalker
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              172
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              male
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              <button
                aria-label="remove from favorites button"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
                id=":r0:"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium css-n4bcw2-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
